File: posts/script/hook.git.go
package main

import (
	"flag"
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"strings"
	"time"
)

// go run main.go --host :8088 --path /data/wwwroot --key 123456

var (
	host     string
	hookPath string
	hookKey  string
	hookCmd  string
)

func main() {
	flag.StringVar(&host, "host", "127.0.0.1:8088", "host")
	flag.StringVar(&hookPath, "path", "", "hook path")
	flag.StringVar(&hookKey, "key", "", "hook key")
	flag.StringVar(&hookCmd, "cmd", "/bin/sh", "sh command")
	flag.Parse()
	if hookPath == "" {
		fmt.Println("please specified hook path.")
		os.Exit(0)
	}
	http.HandleFunc("/hello", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte("Hello World"))
		return
	})
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		r.ParseForm()
		k, ok := r.Form["k"]
		if hookKey != "" {
			if !ok || strings.Compare(strings.Join(k, ""), hookKey) != 0 {
				return
			}
		}
		str := fmt.Sprintf("cd %s && git checkout master && git reset --hard HEAD && git pull", hookPath)
		println("[" + time.Now().Format("2006-01-02 15:04:05") + "] " + str)
		go func() {
			out, err := exec.Command(hookCmd, "-c", str).Output()
			if err != nil {
				println(err.Error())
				return
			}
			println(string(out))
			return
		}()
		fmt.Fprintln(w, "Finish.")
	})
	fmt.Println("The hook server is running at http://" + host)
	fmt.Println("Quit the server with Control-C")
	http.ListenAndServe(host, nil)
}

File: posts/script/hook.git.py
#!/usr/bin/env python
# coding=utf-8

# start:server.py --port=81 --log-file-prefix=/data/www/mgr/log/tornado.81.log

import os
import tornado.web
import tornado.ioloop
import tornado.options
import tornado.httpserver
from tornado.options import define, options

class DefaultHandler(tornado.web.RequestHandler) :
    def get(self):
        self.post()

    def post(self):
        try :
            project = self.get_argument("project")
            path = "/data/www/%s" % (project,)
            if not os.path.exists(path) :
                return self.write("Error");
            if project == 'asm_api' :
                os.system("cd %s && git checkout 4.0.0 && git reset --hard HEAD && git pull" % (path,))
            else :
                os.system("cd %s && git checkout dev && git reset --hard HEAD && git pull" % (path,))
        except :
            return self.write("Hello World")

router = [
    (r'/', DefaultHandler),
]

define("port", default = 8011, help = "run on the given port", type = int)

setting = {
    'debug' : True
}

def main():
    tornado.options.parse_command_line()
    http_server = tornado.httpserver.HTTPServer(tornado.web.Application(
        handlers = router,
        **setting
    ))
    http_server.listen(options.port)

    print("Development server is running at http://127.0.0.1:%s" % options.port)
    print("Quit the server with Control-C")

    tornado.ioloop.IOLoop.instance().start()

if __name__ == "__main__":
    main()
